homeassistant/components/forked_daapd/media_player.py

```diff
@@ -0,0 +1,673 @@
+                if saved_queue:
+                    uris = ""
+                    for item in saved_queue["items"]:
+                        uris += item["uri"] + ","
+                    await self._api.add_to_queue(
+                        uris=uris,
+                        playback="start",
+                        playback_from_position=saved_queue_position,
+                    )
+                    await self._api.seek(position_ms=saved_song_position)
+                    if saved_state != STATE_PLAYING:
+                        await self.async_media_pause()
+        else:
+            _LOGGER.debug("Media type '%s' not supported.", media_type)
+
+
+class ForkedDaapdUpdater:
+    """Manage updates for the forked-daapd device."""
+
+    def __init__(self, hass, api):
+        """Initialize."""
+        self.hass = hass
+        self._api = api
+        self.websocket_handler = None
+        self._all_output_ids = []
+
+    async def async_init(self):
+        """Perform async portion of class initialization."""
+        server_config = await self._api.get_request("config")
+        websocket_port = server_config.get("websocket_port")
+        if websocket_port:
+            self.websocket_handler = asyncio.create_task(
+                self._api.start_websocket_handler(
+                    server_config["websocket_port"],
+                    WS_NOTIFY_EVENT_TYPES,
+                    self._update,
+                    WEBSOCKET_RECONNECT_TIME,
+                    self._disconnected_callback,
+                )
+            )
+        else:
+            _LOGGER.error("Invalid websocket port.")
+
+    def _disconnected_callback(self):
+        async_dispatcher_send(self.hass, SIGNAL_UPDATE_MASTER, False)
+        async_dispatcher_send(self.hass, SIGNAL_UPDATE_OUTPUTS, [])
+
+    async def _update(self, update_types):
+        """Private update method."""
+
+        def intersect(list1, list2):  # local helper
+            intersection = [i for i in list1 + list2 if i in list1 and i in list2]
+            return bool(intersection)
+
+        _LOGGER.debug("Updating %s", update_types)
+        if "queue" in update_types:  # update queue
+            queue = await self._api.get_request("queue")
+            async_dispatcher_send(self.hass, SIGNAL_UPDATE_QUEUE, queue)
+            await self.hass.async_block_till_done()  # make sure queue done before player for async_play_media
+        # order of below don't matter
+        if intersect(["player", "options", "volume"], update_types):  # update player
+            player = await self._api.get_request("player")
+            async_dispatcher_send(self.hass, SIGNAL_UPDATE_PLAYER, player)
+        if intersect(["outputs", "volume"], update_types):  # update outputs
+            outputs = (await self._api.get_request("outputs"))["outputs"]
+            await self._add_zones(outputs)
+            async_dispatcher_send(self.hass, SIGNAL_UPDATE_OUTPUTS, outputs)
```

Comment:
Since we already gave the new entities the latest outputs, we shouldn't need to signal those again to update. We can just have them update state based on the output they already have. We only need to signal the old entities to update. This way we don't need to wait until the new entities have been added before we signal the old entities to update.

Reply:
Sure.
With respect to entity lifecycle, can we assume users cannot delete the entities manually? As I understand it, the integration lifecycle begins with async_setup_entry and ends with the async_unload_entry. Entities only get permanently removed on async_unload_entry then too (but they can get occasional calls to async_added_to_hass and async_will_remove_from_hass, eg when renaming). Is this correct?

Reply:
Yes, the entity will be removed and re-added if the user changes the entity_id. Also if the user disables an entity, the entity will be removed, or enables a disabled entity, the entity will be added again.